README.md
# Rari Stable Pool: Smart Contracts

Welcome to `rari-stable-pool-contracts`, the central repository for the Solidity source code behind the Rari Stable Pool's Ethereum-based smart contracts (with automated tests and documentation).

## How it works

The Rari Stable Pool is a decentralized and fully-audited stablecoin lending aggregator optimized for yield based on the Ethereum blockchain. A high-level overview of how the Rari Stable Pool works is available in [`CONCEPT.md`](CONCEPT.md). This information is also [available online](https://rari.capital/current.html). Find out more about Rari Capital at [rari.capital](https://rari.capital).

## Contract usage

Documentation on common usage of the contracts is available in [`USAGE.md`](USAGE.md). Detailed API documentation for our smart contracts' public methods is available in [`API.md`](API.md). Smart contract ABIs are available in the `abi` properties of the JSON files in the `build` folder. For easy implementation, see the [Rari JavaScript SDK](https://github.com/Rari-Capital/rari-sdk).

## Installation (for development and deployment)

We, as well as others, had success using Truffle on Node.js `v12.18.2` with the latest version of NPM.

To install the latest version of Truffle: `npm install -g truffle`

*Though the latest version of Truffle should work, to compile, deploy, and test our contracts, we used Truffle `v5.1.45` (which should use `solc` version `0.5.17+commit.d19bba13.Emscripten.clang` and Web3.js `v1.2.1`).*

To install all our dependencies: `npm install`

## Compiling the contracts

`npm run compile`

## Testing the contracts

In `.env`, set `DEVELOPMENT_ADDRESS=0x45D54B22582c79c8Fb8f4c4F2663ef54944f397a` to test deployment and also set `DEVELOPMENT_ADDRESS_SECONDARY=0x1Eeb75CFad36EDb6C996f7809f30952B0CA0B5B9` to run automated tests.

If you are upgrading from `v2.5.0`, set `UPGRADE_FROM_LAST_VERSION=1` to enable upgrading and configure the following:

    UPGRADE_OLD_FUND_CONTROLLER=0x369855b051d1b2dbee88a792dcfc08614ff4e262
    UPGRADE_FUND_MANAGER_ADDRESS=0xC6BF8C8A55f77686720E0a88e2Fd1fEEF58ddf4a
    UPGRADE_FUND_OWNER_ADDRESS=0x10dB6Bce3F2AE1589ec91A872213DAE59697967a

Then, copy the OpenZeppelin artifacts for the official deployed `v2.5.0` contracts from `.openzeppelin/mainnet.json` to `.openzeppelin/unknown-1337.json`. If you decide to disable upgrading by setting restoring `UPGRADE_FROM_LAST_VERSION=0`, make sure to delete `.openzeppelin/unknown-1337.json`.

To test the contracts, first fork the Ethereum mainnet. Begin by configuring `DEVELOPMENT_WEB3_PROVIDER_URL_TO_BE_FORKED` in `.env` (set to any mainnet Web3 HTTP provider JSON-RPC URL; we use a local `geth` instance, specifically a light client started with `geth --syncmode light --rpc --rpcapi eth,web3,debug,net`; Infura works too, but beware of latency and rate limiting). To start the fork, run `npm run ganache`. *If you would like to change the port, make sure to configure `scripts/ganache.js`, `scripts/test.sh`, and the `development` network in `truffle-config.js`.* Note that you will likely have to regularly restart your fork, especially when forking from a node without archive data or when using live 0x API responses to make currency exchanges.

To deploy the contracts to your private mainnet fork: `truffle migrate --network development --skip-dry-run --reset`

To run automated tests on the contracts on your private mainnet fork, run `npm test` (which runs `npm run ganache` in the background for you). If you are upgrading from `v2.5.0`, you must also set the following variables in `.env`:

    UPGRADE_FUND_TOKEN_ADDRESS=0x016bf078ABcaCB987f0589a6d3BEAdD4316922B0
    UPGRADE_FUND_PRICE_CONSUMER_ADDRESS=0xFE98A52bCAcC86432E7aa76376751DcFAB202244

If you'd like to test gasless deposits via `RariFundProxy.deposit` via the Gas Station Network:

* Download `https://github.com/OpenZeppelin/openzeppelin-gsn-provider/blob/master/bin/gsn-relay` to `bin/gsn-relay` and set permissions with `chmod +x bin/gsn-relay`.
* Making sure `npx` is installed, run `npm dev-gsn`.
* Fund `RariFundProxy` using `npx @openzeppelin/gsn-helpers fund-recipient --recipient $RARI_FUND_PROXY_ADDRESS -n http://localhost:8546 -f $FROM_ADDRESS` or [this tool](https://www.opengsn.org/recipients) (or manually send ETH to `RelayHub(0xD216153c06E857cD7f72665E0aF1d7D82172F494).depositFor(address target)`).
* Run `rari-gsn-signer` with `pm2 start ecosystem.config.js` after configuring `ecosystem.config.js`.

## Live deployment

In `.env`, configure `LIVE_DEPLOYER_ADDRESS`, `LIVE_DEPLOYER_PRIVATE_KEY`, `LIVE_WEB3_PROVIDER_URL`, `LIVE_GAS_PRICE` (ideally, use the "fast" price listed by [ETH Gas Station](https://www.ethgasstation.info/)), `LIVE_FUND_OWNER`, `LIVE_FUND_REBALANCER`, `LIVE_FUND_INTEREST_FEE_MASTER_BENEFICIARY`, `LIVE_FUND_WITHDRAWAL_FEE_MASTER_BENEFICIARY`, and `LIVE_FUND_GSN_TRUSTED_SIGNER` to deploy to the mainnet.

If you are upgrading from `v2.5.0`, set `UPGRADE_FROM_LAST_VERSION=1` to enable upgrading and configure the following:

    UPGRADE_OLD_FUND_CONTROLLER=0x369855b051d1b2dbee88a792dcfc08614ff4e262
    UPGRADE_FUND_MANAGER_ADDRESS=0xC6BF8C8A55f77686720E0a88e2Fd1fEEF58ddf4a
    UPGRADE_FUND_OWNER_ADDRESS=0x10dB6Bce3F2AE1589ec91A872213DAE59697967a

You must also set `LIVE_UPGRADE_FUND_OWNER_PRIVATE_KEY`.

Then, migrate: `truffle migrate --network live`

If you'd like to provide gasless deposits via `RariFundProxy.deposit` via the Gas Station Network:

* Fund `RariFundProxy` using `npx @openzeppelin/gsn-helpers fund-recipient --recipient $RARI_FUND_PROXY_ADDRESS -n $ETHEREUM_NODE_URL -f $FROM_ADDRESS` or [this tool](https://www.opengsn.org/recipients) (or manually send ETH to `RelayHub(0xD216153c06E857cD7f72665E0aF1d7D82172F494).depositFor(address target)`).
* Run `rari-gsn-signer` with `pm2 start ecosystem.config.js --env production` after configuring `ecosystem.config.js`.

## License

See `LICENSE`.

## Credits

Rari Capital's smart contracts are developed by [David Lucid](https://github.com/davidlucid). Find out more about Rari Capital at [rari.capital](https://rari.capital).


CHANGELOG.md
# Changelog

## `v2.6.0` (contracts deployed 2021-05-04; all code pushed 2021-05-06)

* Fuse integration!
* Upgrade to mStable mUSD v3.
* Integrate swaps via Uniswap V2 in `RariFundController`.

## `v2.5.0` (contracts deployed 2021-02-09; all code pushed 2021-02-14)

* Upgraded mStable SAVE V1 to V2 (and implemented MTA rewards).
* Fixed bug in calculation of `outputFilledAmountUsd` in `RariFundController.marketSell0xOrdersFillOrKill`.
* Check `fundDisabled` in `RariFundManager.upgradeFundController`.

## `v2.4.1` (no contracts deployed; all code pushed 2020-12-07)

* Updated mStable swap tests to skip removed bAssets (i.e., DAI).

## `v2.4.0` (contracts deployed 2020-11-25; all code pushed 2020-11-29)

* Fixed bug in validation of mUSD redemption in `RariFundController.withdrawAndExchange`.

## `v2.3.0` (contracts deployed 2020-10-29; all code pushed 2020-11-13)

* Added `RariFundPriceConsumer.allCurrenciesPeggedTo1Usd` variable.

## `v2.2.0` (contracts deployed 2020-10-27; all code pushed 2020-10-29)

* Added 0.5% withdrawal fee.
    * Removed `RariFundManager.withdrawFees` to save gas on deployment.

## `v2.1.0` (contracts deployed 2020-10-20; all code pushed 2020-10-28)

* Implemented liquidity mining of RGT (Rari Governance Token) distributions.
* Removed account balance limit functions.
* Moved dApp to `rari-dapp-legacy`.
* Gave `RariFundManager` permission to burn RSPT on withdrawals without the need for approval transactions.
* Various minor improvements to `RariFundManager` (e.g., replaced `disableFund` and `enableFund` with `setFundDisabled`).
* Fixed `LIVE_UPGRADE_FUND_OWNER_PRIVATE_KEY` in `truffle-config.js`.
* Changed performance/interest fee rate to 9.5%.

## `v2.0.0` (contracts deployed 2020-09-21; all code pushed 2020-10-05)

* Rebranded to Rari Stable Pool and Rari Stable Pool Token (RSPT).
* Implemented lending via Aave.
* Implemented lending and exchanges via mStable.
* Removed account balance limit.
* Implemented daily limit on proportion of funds lost due to exchange.
* Added currency and pool allocation pie chart to web client.
* Fixed slippage calculation logic in web client.
* Fixed Quantstamp audit findings:
    * Implement stablecoin pricing via Chainlink (QSP-1).
    * Replaced calls to `transfer()` with `call.value()` (QSP-3).
    * Implemented proxy storage for `RariFundManager`, `RariFundToken`, and `RariFundPriceConsumer` (QSP-3).
    * Upgrade RFT, removing `interestAccruedBy` to save gas and remove `initNetDeposits` (QSP-3).
    * Replace imitation of `balanceOfUnderlying` with actual call to `balanceOfUnderlying` in `CompoundPoolController` (QSP-4).
    * Make sure new contract address is a `RariFundController` in `upgradeFundController` (QSP-5).
    * Updated docs (QSP-7, QSP-10).
    * Make sure sender is the 0x Exchange v3 contract in fallback functions of `RariFundController` and `RariFundProxy` (QSP-8).
    * Check for overflow before conversion from uint256 to int256 (QSP-9).
    * Locked all Solidity version pragmas to 0.5.17 (QSP-11, QSP-12).
    * Improved code comments (QSP-10, QSP-14).
* Fixed bug in which `RariFundManager._withdrawFrom` could have overestimated the amount of RFT burned (see commit [9059e8c](https://github.com/Rari-Capital/rari-stable-pool-contracts/commit/9059e8c4c8c9de545680b70f20981f261bfc425d)).
    * It seems that this bug did not cause any issues before it was fixed: it was discovered during development and was never reported by any users.
* Other changes to contracts:
    * Externalized pool and exchange libraries.
    * Added events for pool allocations.
    * Optimized `RariFundController`.
    * Add `RariFundManager.getAcceptedCurrencies` and replace `RariFundController.getPoolBalances` with `RariFundProxy.getRawFundBalancesAndPrices` to speed up web client.
    * Replaced `setAcceptedCurrency` with `setAcceptedCurrencies` to minimize rebalancer gas usage.
    * Added `getAccountBalanceLimit` function to `RariFundManager`.
    * Moved token forwarding logic from `RariFundManager.setFundController` to `RariFundController.upgradeFundController`.
    * Store decimal precision of tokens in contracts instead of querying the token contracts (to save gas).
    * Added `forwardLostFunds` function to RariFundManager and RariFundProxy to forward lost tokens accidentally transferred directly to these contracts.
    * Made Rari contract variables public in contracts.
    * Removed unnecessary boolean return values.
    * Removed version constants from all contracts.
    * Add fund owner as minter of RSPT in case of emergency.
    * Other small improvements, bug fixes, comments, documentation, etc.
* Other changes to testing:
    * Added Travis CI.
        * Run Ganache beforehand in test script so Travis CI works.
    * Unlock development accounts in Ganache so we don't need private keys for testing.
    * Reordered tests so 0x exchanges come first so orders from 0x API are fresh.
    * Assert with margin of error in `test/3_fund_owner.js`.
    * Fixed accepted margin of error in `test/5_fund_user.js`.
    * Added test to check gas used by `RariFundController.upgradeFundController`.
    * Added test to check gas used by `RariFundProxy.withdrawAndExchange`.
    * Moved scripts to `scripts` folder and set as executable.
    * Increased block gas limit to 12.5 million in `scripts/ganache.js`.
    * Minor improvements to test dummy contracts.
    * Other improvements to test scripts.
    * Added copyright notices to tests.
* Other changes to web client:
    * Updated web3.js and providers used by web client.
    * Other minor web client improvements and bug fixes.
* Other changes to documentation:
    * Added `USAGE.md`.
    * Improved `README.md`, `API.md`, and `DEPLOYED.md`.
    * Renamed `FUND.md` to `CONCEPT.md` and improved content.
    * Added UML class diagrams for contracts.
    * Added Solidity metrics report.
* Other changes to package:
    * Replaced patched fork of `ganche-core` with genuine beta update.

migrations/2_deploy_contracts.js
// SPDX-License-Identifier: UNLICENSED
const { deployProxy, upgradeProxy, admin } = require('@openzeppelin/truffle-upgrades');
require('dotenv').config();

var DydxPoolController = artifacts.require("./lib/pools/DydxPoolController.sol");
var CompoundPoolController = artifacts.require("./lib/pools/CompoundPoolController.sol");
var AavePoolController = artifacts.require("./lib/pools/AavePoolController.sol");
var MStablePoolController = artifacts.require("./lib/pools/MStablePoolController.sol");
var FusePoolController = artifacts.require("./lib/pools/FusePoolController.sol");
var ZeroExExchangeController = artifacts.require("./lib/exchanges/ZeroExExchangeController.sol");
var MStableExchangeController = artifacts.require("./lib/exchanges/MStableExchangeController.sol");
var UniswapExchangeController = artifacts.require("./lib/exchanges/UniswapExchangeController.sol");
var RariFundController = artifacts.require("./RariFundController.sol");
var RariFundManager = artifacts.require("./RariFundManager.sol");
var RariFundToken = artifacts.require("./RariFundToken.sol");
var RariFundPriceConsumer = artifacts.require("./RariFundPriceConsumer.sol");
var RariFundProxy = artifacts.require("./RariFundProxy.sol");

module.exports = async function(deployer, network, accounts) {
  if (["live", "live-fork"].indexOf(network) >= 0) {
    if (!process.env.LIVE_GAS_PRICE) return console.error("LIVE_GAS_PRICE is missing for live deployment");
    if (!process.env.LIVE_FUND_OWNER) return console.error("LIVE_FUND_OWNER is missing for live deployment");
    if (!process.env.LIVE_FUND_REBALANCER) return console.error("LIVE_FUND_REBALANCER is missing for live deployment");
    if (!process.env.LIVE_FUND_INTEREST_FEE_MASTER_BENEFICIARY) return console.error("LIVE_FUND_INTEREST_FEE_MASTER_BENEFICIARY is missing for live deployment");
    if (!process.env.LIVE_FUND_WITHDRAWAL_FEE_MASTER_BENEFICIARY) return console.error("LIVE_FUND_WITHDRAWAL_FEE_MASTER_BENEFICIARY is missing for live deployment");
  }
  
  if (parseInt(process.env.UPGRADE_FROM_LAST_VERSION) > 0) {
    // Upgrade from v2.4.0 (only modifying RariFundController v2.0.0 and RariFundManager v2.2.0) to v2.5.0
    if (!process.env.UPGRADE_OLD_FUND_CONTROLLER) return console.error("UPGRADE_OLD_FUND_CONTROLLER is missing for upgrade");
    if (!process.env.UPGRADE_FUND_MANAGER_ADDRESS) return console.error("UPGRADE_FUND_MANAGER_ADDRESS is missing for upgrade");
    if (!process.env.UPGRADE_FUND_OWNER_ADDRESS) return console.error("UPGRADE_FUND_OWNER_ADDRESS is missing for upgrade");

    if (["live", "live-fork"].indexOf(network) >= 0) {
      if (!process.env.LIVE_UPGRADE_FUND_OWNER_PRIVATE_KEY) return console.error("LIVE_UPGRADE_FUND_OWNER_PRIVATE_KEY is missing for live upgrade");
    } else {
      if (!process.env.UPGRADE_FUND_TOKEN_ADDRESS) return console.error("UPGRADE_FUND_TOKEN_ADDRESS is missing for development upgrade");
      if (!process.env.UPGRADE_FUND_PRICE_CONSUMER_ADDRESS) return console.error("UPGRADE_FUND_PRICE_CONSUMER_ADDRESS is missing for development upgrade");
    }

    // Upgrade from v2.4.0 (RariFundController v2.0.0) to v2.5.0
    var oldRariFundController = await RariFundController.at(process.env.UPGRADE_OLD_FUND_CONTROLLER);

    // Deploy liquidity pool and currency exchange libraries
    await deployer.deploy(DydxPoolController);
    await deployer.deploy(CompoundPoolController);
    await deployer.deploy(AavePoolController);
    await deployer.deploy(MStablePoolController);
    await deployer.deploy(FusePoolController);
    await deployer.deploy(MStableExchangeController);
    await deployer.deploy(UniswapExchangeController);

    // Link libraries to RariFundController
    await deployer.link(DydxPoolController, RariFundController);
    await deployer.link(CompoundPoolController, RariFundController);
    await deployer.link(AavePoolController, RariFundController);
    await deployer.link(MStablePoolController, RariFundController);
    await deployer.link(FusePoolController, RariFundController);
    await deployer.link(MStableExchangeController, RariFundController);
    await deployer.link(UniswapExchangeController, RariFundController);

    // Deploy new RariFundController
    var rariFundController = await deployProxy(RariFundController, { deployer, unsafeAllowLinkedLibraries: true });

    // Disable the fund on the old RariFundController
    await oldRariFundController.disableFund({ from: process.env.UPGRADE_FUND_OWNER_ADDRESS });

    // Disable the fund on the RariFundManager
    RariFundManager.class_defaults.from = process.env.UPGRADE_FUND_OWNER_ADDRESS;
    var rariFundManager = await RariFundManager.at(process.env.UPGRADE_FUND_MANAGER_ADDRESS);
    await rariFundManager.setFundDisabled(true);

    // Upgrade RariFundController
    await oldRariFundController.methods["upgradeFundController(address)"](RariFundController.address, { from: process.env.UPGRADE_FUND_OWNER_ADDRESS });

    // Forward COMP governance tokens
    if (oldRariFundController.methods["upgradeFundController(address,address)"].call(RariFundController.address, "0xc00e94cb662c3520282e6f5717214004a7f26888", { from: process.env.UPGRADE_FUND_OWNER_ADDRESS })) await oldRariFundController.methods["upgradeFundController(address,address)"](RariFundController.address, "0xc00e94cb662c3520282e6f5717214004a7f26888", { from: process.env.UPGRADE_FUND_OWNER_ADDRESS });

    // Connect new RariFundController and RariFundManager
    await rariFundController.setFundManager(process.env.UPGRADE_FUND_MANAGER_ADDRESS);
    await rariFundManager.setFundController(RariFundController.address);

    // Upgrade RariFundManager
    rariFundManager = await upgradeProxy(process.env.UPGRADE_FUND_MANAGER_ADDRESS, RariFundManager, { deployer, unsafeAllowCustomTypes: true });

    // Re-enable the fund on the RariFundManager
    await rariFundManager.setFundDisabled(false);

    // Set Aave referral code
    await rariFundController.setAaveReferralCode(86);

    // Set daily loss rate limit for currency exchanges
    await rariFundController.setExchangeLossRateLimit(["live", "live-fork"].indexOf(network) >= 0 ? web3.utils.toBN(2).pow(web3.utils.toBN(255)).neg() : web3.utils.toBN(0.5e18));

    // Set fund rebalancer on controller and manager
    await rariFundController.setFundRebalancer(["live", "live-fork"].indexOf(network) >= 0 ? process.env.LIVE_FUND_REBALANCER : process.env.DEVELOPMENT_ADDRESS);

    // Deploy ZeroExExchangeController for RariFundProxy
    await deployer.deploy(ZeroExExchangeController);

    // Link libraries to RariFundProxy
    await deployer.link(ZeroExExchangeController, RariFundProxy);
    await deployer.link(MStableExchangeController, RariFundProxy);

    // Deploy RariFundProxy
    var rariFundProxy = await deployer.deploy(RariFundProxy);

    // Connect RariFundManager and RariFundProxy
    await rariFundManager.setFundProxy(RariFundProxy.address);
    await rariFundProxy.setFundManager(RariFundManager.address);

    // Set GSN trusted signer
    await rariFundProxy.setGsnTrustedSigner(["live", "live-fork"].indexOf(network) >= 0 ? process.env.LIVE_FUND_GSN_TRUSTED_SIGNER : process.env.DEVELOPMENT_ADDRESS);

    if (["live", "live-fork"].indexOf(network) >= 0) {
      // Live network: transfer ownership of RariFundController to live owner
      await rariFundController.transferOwnership(process.env.LIVE_FUND_OWNER);
      await rariFundProxy.transferOwnership(process.env.LIVE_FUND_OWNER);
    } else {
      // Register Fuse pools
      var testFusePools = require("../test/fixtures/fuse.json");
      var poolKeys = Object.keys(testFusePools);
      var poolIds = [];
      var currencyCodes = [];
      var cTokens = [];
      for (var i = 0; i < poolKeys.length; i++) {
        poolIds[i] = 100 + i;
        currencyCodes[i] = [];
        cTokens[i] = [];
        for (const currencyCode of Object.keys(testFusePools[poolKeys[i]].currencies)) {
          currencyCodes[i].push(currencyCode);
          cTokens[i].push(testFusePools[poolKeys[i]].currencies[currencyCode].cTokenAddress);
        }
      }
      await rariFundController.addFuseAssets(poolIds, currencyCodes, cTokens);

      // Development network: transfer ownership of contracts to development address, set development address as rebalancer, and set all currencies to accepted
      await rariFundManager.transferOwnership(process.env.DEVELOPMENT_ADDRESS, { from: process.env.UPGRADE_FUND_OWNER_ADDRESS });
      var rariFundPriceConsumer = await RariFundPriceConsumer.at(process.env.UPGRADE_FUND_PRICE_CONSUMER_ADDRESS); 
      await rariFundPriceConsumer.transferOwnership(process.env.DEVELOPMENT_ADDRESS, { from: process.env.UPGRADE_FUND_OWNER_ADDRESS });
      // TODO: await admin.transferProxyAdminOwnership(process.env.DEVELOPMENT_ADDRESS, { from: process.env.UPGRADE_FUND_OWNER_ADDRESS });
      RariFundManager.class_defaults.from = process.env.DEVELOPMENT_ADDRESS;
      await rariFundManager.setFundRebalancer(process.env.DEVELOPMENT_ADDRESS);
      await rariFundManager.setAcceptedCurrencies(["DAI", "USDC", "USDT", "TUSD", "BUSD", "sUSD", "mUSD"], [true, true, true, true, true, true, true]);
    }
  } else {
    // Normal deployment!
    // Deploy liquidity pool and currency exchange libraries
    await deployer.deploy(DydxPoolController);
    await deployer.deploy(CompoundPoolController);
    await deployer.deploy(AavePoolController);
    await deployer.deploy(MStablePoolController);
    await deployer.deploy(FusePoolController);
    await deployer.deploy(MStableExchangeController);
    await deployer.deploy(UniswapExchangeController);

    // Link libraries to RariFundController
    await deployer.link(DydxPoolController, RariFundController);
    await deployer.link(CompoundPoolController, RariFundController);
    await deployer.link(AavePoolController, RariFundController);
    await deployer.link(MStablePoolController, RariFundController);
    await deployer.link(FusePoolController, RariFundController);
    await deployer.link(MStableExchangeController, RariFundController);
    await deployer.link(UniswapExchangeController, RariFundController);

    // Deploy RariFundController and RariFundManager
    var rariFundController = await deployProxy(RariFundController, { deployer, unsafeAllowLinkedLibraries: true });
    var rariFundManager = await deployProxy(RariFundManager, [], { deployer });

    // Connect RariFundController and RariFundManager
    await rariFundController.setFundManager(RariFundManager.address);
    await rariFundManager.setFundController(RariFundController.address);

    // Set Aave referral code
    await rariFundController.setAaveReferralCode(86);
    
    // Deploy RariFundToken
    var rariFundToken = await deployProxy(RariFundToken, [], { deployer });
    
    // Add RariFundManager as as RariFundToken minter
    await rariFundToken.addMinter(RariFundManager.address);

    // Connect RariFundToken to RariFundManager
    await rariFundManager.setFundToken(RariFundToken.address);

    // Deploy RariFundPriceConsumer, pegging all currencies to $1 USD
    var rariFundPriceConsumer = await deployProxy(RariFundPriceConsumer, [true], { deployer });

    // Connect RariFundPriceConsumer to RariFundManager
    await rariFundManager.setFundPriceConsumer(RariFundPriceConsumer.address);

    // Set daily loss rate limit for currency exchanges
    await rariFundController.setExchangeLossRateLimit(["live", "live-fork"].indexOf(network) >= 0 ? web3.utils.toBN(2).pow(web3.utils.toBN(255)).neg() : web3.utils.toBN(0.5e18));

    // Set fund rebalancer on controller and manager
    await rariFundController.setFundRebalancer(["live", "live-fork"].indexOf(network) >= 0 ? process.env.LIVE_FUND_REBALANCER : process.env.DEVELOPMENT_ADDRESS);
    await rariFundManager.setFundRebalancer(["live", "live-fork"].indexOf(network) >= 0 ? process.env.LIVE_FUND_REBALANCER : process.env.DEVELOPMENT_ADDRESS);

    // Set interest fee master beneficiary
    await rariFundManager.setInterestFeeMasterBeneficiary(["live", "live-fork"].indexOf(network) >= 0 ? process.env.LIVE_FUND_INTEREST_FEE_MASTER_BENEFICIARY : process.env.DEVELOPMENT_ADDRESS);

    // Set interest fee rate to 9.5%
    await rariFundManager.setInterestFeeRate(web3.utils.toBN(0.095e18));
  
    // Set withdrawal fee master beneficiary
    await rariFundManager.setWithdrawalFeeMasterBeneficiary(["live", "live-fork"].indexOf(network) >= 0 ? process.env.LIVE_FUND_WITHDRAWAL_FEE_MASTER_BENEFICIARY : process.env.DEVELOPMENT_ADDRESS);
  
    // Set withdrawal fee rate to 0.5%
    await rariFundManager.setWithdrawalFeeRate(web3.utils.toBN(0.005e18));

    // Deploy ZeroExExchangeController for RariFundProxy
    await deployer.deploy(ZeroExExchangeController);

    // Link libraries to RariFundProxy
    await deployer.link(ZeroExExchangeController, RariFundProxy);
    await deployer.link(MStableExchangeController, RariFundProxy);

    // Deploy RariFundProxy
    var rariFundProxy = await deployer.deploy(RariFundProxy);

    // Connect RariFundManager and RariFundProxy
    await rariFundManager.setFundProxy(RariFundProxy.address);
    await rariFundProxy.setFundManager(RariFundManager.address);

    // Set GSN trusted signer
    await rariFundProxy.setGsnTrustedSigner(["live", "live-fork"].indexOf(network) >= 0 ? process.env.LIVE_FUND_GSN_TRUSTED_SIGNER : process.env.DEVELOPMENT_ADDRESS);

    if (["live", "live-fork"].indexOf(network) >= 0) {
      // Live network: transfer ownership of deployed contracts from the deployer to the owner
      await rariFundController.transferOwnership(process.env.LIVE_FUND_OWNER);
      await rariFundManager.transferOwnership(process.env.LIVE_FUND_OWNER);
      await rariFundToken.addMinter(process.env.LIVE_FUND_OWNER);
      await rariFundToken.renounceMinter();
      await rariFundProxy.transferOwnership(process.env.LIVE_FUND_OWNER);
      await rariFundPriceConsumer.transferOwnership(process.env.LIVE_FUND_OWNER);
      await admin.transferProxyAdminOwnership(process.env.LIVE_FUND_OWNER);
    } else {
      // Register Fuse pools
      var testFusePools = require("../test/fixtures/fuse.json");
      var poolKeys = Object.keys(testFusePools);
      var poolIds = [];
      var currencyCodes = [];
      var cTokens = [];
      for (var i = 0; i < poolKeys.length; i++) {
        poolIds[i] = 100 + i;
        currencyCodes[i] = [];
        cTokens[i] = [];
        for (const currencyCode of Object.keys(testFusePools[poolKeys[i]].currencies)) {
          currencyCodes[i].push(currencyCode);
          cTokens[i].push(testFusePools[poolKeys[i]].currencies[currencyCode].cTokenAddress);
        }
      }
      await rariFundController.addFuseAssets(poolIds, currencyCodes, cTokens);

      // Development network: set all currencies to accepted
      await rariFundManager.setAcceptedCurrencies(["DAI", "USDC", "USDT", "TUSD", "BUSD", "sUSD", "mUSD"], [true, true, true, true, true, true, true]);
    }
  }
};
